src/base/PolymeshTransactionBase.ts

```diff
@@ -293,80 +362,132 @@ export abstract class PolymeshTransactionBase<Values extends unknown[] = unknown
    *
-   * @note this value might change if the transaction is run at a later time. This can be due to a governance vote
+   * @param listener - callback function that will be called whenever the middleware is updated with the latest data.
+   *   If there is an error (timeout or middleware offline) it will be passed to this callback
+   *
+   * @note this event will be fired even if the queue fails
+   * @returns unsubscribe function
+   * @throws if the middleware wasn't enabled when instantiating the SDK client
    */
-  public async getFees(): Promise<Fees | null> {
-    const { signingAddress } = this;
-    let { protocolFee: protocol } = this;
+  public onProcessedByMiddleware(listener: (err?: PolymeshError) => void): () => void {
+    if (!this.context.isMiddlewareEnabled()) {
+      throw new PolymeshError({
+        code: ErrorCode.General,
+        message: 'Cannot subscribe without an enabled middleware connection',
+      });
+    }
 
-    let composedTx;
+    this.emitter.on(Event.ProcessedByMiddleware, listener);
 
-    try {
-      composedTx = this.composeTx();
-    } catch (err) {
-      return null;
-    }
+    return (): void => {
+      this.emitter.removeListener(Event.ProcessedByMiddleware, listener);
+    };
+  }
 
-    const paymentInfoPromise = composedTx.paymentInfo(signingAddress);
+  /**
+   * Poll the middleware every 2 seconds to see if it has already processed the
+   *   block that reflects the changes brought on by this transaction being run. If so,
+   *   emit the corresponding event. After 5 retries (or if the middleware can't be reached),
+   *   the event is emitted with an error
+   *
+   * @note uses the middleware
+   */
+  private async emitWhenMiddlewareIsSynced(): Promise<void> {
+    const { context } = this;
 
-    if (!protocol) {
-      protocol = await this.getProtocolFees();
+    if (!context.isMiddlewareEnabled()) {
+      return;
     }
 
-    const { partialFee } = await paymentInfoPromise;
+    const blockNumber = await context.getLatestBlock();
 
-    return {
-      protocol,
-      gas: balanceToBigNumber(partialFee),
-    };
+    let done = false;
+
+    /*
+     * We do not await this promise because it is supposed to run in the background, and
+     * any errors encountered are emitted. If the user isn't listening, they shouldn't
+     * care about middleware (or other) errors anyway
+     */
+    P.each(range(6), async i => {
+      if (done) {
+        return;
+      }
+
+      try {
+        const {
+          data: {
+            latestBlock: { id: processedBlock },
+          },
+        } = await context.queryMiddleware<Ensured<Query, 'latestBlock'>>(latestProcessedBlock());
+
+        if (blockNumber.lte(processedBlock)) {
+          done = true;
+          this.emitter.emit(Event.ProcessedByMiddleware);
+          return;
+        }
+      } catch (err) {}
```

Comment:
Do we need this catch here? Wouldn't the `.catch` handle this error if it does happen? If there is a certain type of error this is meant, can it be scoped to the type at least.

I think its OK to leave in (its only one query), but generally I think swallowing every error silently can be frustrating to debug if there is an issue.

Reply:
We swallow the error because if there's a problem running the query we just want to retry, not have the entire process fail

Reply:
I'll add a comment explaining this